Added Python Documentation (#128)
spokestack/spokestack-website: content/docs/Python/getting-started.md

---
title: Getting Started
navId: Getting Started (Python)
description: Getting started with the Spokestack Python API
draft: false
---

This guide will get you up and running with Spokestack for Python, and you'll have a voice interface in your application in no time.

## Installation

### System Dependencies

> There are some system dependencies that need to be downloaded in order to install spokestack via pip.

Reviewer: ```suggestion
There are some system dependencies that need to be downloaded in order to install Spokestack via pip.
```

#### macOS

```shell
brew install lame portaudio
```

#### Debian/Ubuntu

```shell
sudo apt-get install portaudio19-dev libmp3lame-dev
```

#### Windows

> We currently do not support Windows 10 natively, and recommend you install [Windows Subsystem for Linux (WSL)](https://docs.microsoft.com/en-us/windows/wsl/install-win10) with the Debian dependencies. However, if you would like to work on native Windows support we will glady accept pull requests.

Reviewer: ```suggestion
We currently do not support Windows 10 natively, and recommend you install [Windows Subsystem for Linux (WSL)](https://docs.microsoft.com/en-us/windows/wsl/install-win10) with the Debian dependencies. However, if you would like to work on native Windows support, we gladly accept pull requests.
```

> Another potential avenue for using `spokestack` on Windows 10 is from [anaconda](https://www.anaconda.com/). This is without support for Text To Speech (TTS) though due to the Lame dependency. PortAudio, on the other hand, can be installed via `conda`.

Reviewer: ```suggestion
Another potential avenue for using Spokestack on Windows 10 is via [anaconda](https://www.anaconda.com/). PortAudio can be installed via `conda`, but Lame cannot. Hence, microphone input will be supported, but text-to-speech will not.
```

I reordered these because it feels odd to read two sentences in a row that introduce an exception to the previous sentence.

```shell
conda install portaudio
```

### Installation with pip

Once system dependencies have been satisfied, you can install the library with the following.

```shell
pip install spokestack
```

### Setup

> We use `pyenv` for virtual environments. Below you will find the step-by-step commands to install a virtual environment.

Reviewer: ```suggestion
We use `pyenv` for virtual environments.
```

```shell
pyenv install 3.8.6
pyenv virtualenv 3.8.6 spokestack
pyenv local spokestack
pip install -r requirements.txt
```

### Install Tensorflow

This library requires a way to run [TFLite](https://www.tensorflow.org/lite) models. There are two ways to add this ability. The first is installing the full [Tensorflow](https://www.tensorflow.org/) library.

> The full Tensorflow package is installed with the following:

Reviewer: ```suggestion
This library requires a way to run [TFLite](https://www.tensorflow.org/lite) models. There are two ways to add this ability. The first is installing the full [Tensorflow](https://www.tensorflow.org/) library:
```

```shell
pip install tensorflow
```

#### TFLite Interpreter (Embedded Devices)

> 

Reviewer: ```suggestion
```

The last section says "The first [way]", but it never discusses the second—this is easier to recognize as the second installation method if it's not separated by a header line.

In use cases where you require a small footprint, such as on a [Raspberry Pi](https://www.raspberrypi.org/) or similar [Internet of Things (IOT)](https://en.wikipedia.org/wiki/Internet_of_things) devices, you will want to install the TFLite Interpreter. You can install it for your platform by [following the instructions](https://www.tensorflow.org/lite/guide/python#install_just_the_tensorflow_lite_interpreter).

## Integration

> In order for your application to use the Spokestack features, there are a few things you will need:

Reviewer: ```suggestion
In order for your application to use Spokestack's features, there are a few things you will need:
```

- A free Spokestack Account
- Audio Input Device
- A `SpeechPipeline` instance

> - Audio Output Device

Reviewer: ```suggestion
- A free Spokestack account
- Audio input device
- A `SpeechPipeline` instance
- Audio output device
```

Reviewer: These should be consistent—either the things I've lowercased should be lowercased, or "instance" should be capitalized (and I'd recommend following the same for the section headers below). I prefer to not overuse capitalization, so I went this way; the other way's probably OK too.

### 1. Spokestack Account

> Go to [spokestack.io](/account/) to set up your own account (it's free!). Once you've got that, go [grab one of our free NLU models](/account/services/nlu). We'll use the `Highlow` one in this example, but you can choose another, or [create your own](/docs/Concepts/nlu-training-data)

Reviewer: ```suggestion
Go to [spokestack.io](/create) to set up your own account (it's free!). Once you've got that, go [grab one of our free NLU models](/account/services/nlu). We'll use the `Highlow` one in this example, but you can choose another, or [create your own](/docs/concepts/nlu-training-data)
```

> Once you've downloaded your NLU, unzip `nlu.tar.gz` and add the directory with the three files inside (`metadata.json`, `nlu.tflite`, `vocab.txt`) to your Python project.

Reviewer: I might add a sentence about how it doesn't matter where they're added since you'll be passing paths to them later on—technically, they don't even need to be "in the project".

### 2. Audio Input Device

The `PyAudioInput` class will use the system default audio input device. Most personal computers have some form of microphone, but in the case of an embedded device, you may need to purchase a small USB microphone.

### 3. `SpeechPipeline` instance

The `SpeechPipeline` [guide](speech-pipeline) has a detailed explanation of how to set up the pipeline, so we will just show an example here.

```python
from spokestack.pipeline import SpeechPipeline
from spokestack.io.pyaudio import PyAudioInput
from spokestack.vad.webrtc import VoiceActivityDetector
from spokestack.wakeword.tflite import WakewordTrigger
from spokestack.asr.spokestack.speech_recognizer import SpeechRecognizer

pipeline = SpeechPipeline(
    PyAudioInput(),
    [
        VoiceActivityDetector(),

>         WakewordTrigger("path_to_tflite_model"),

Reviewer 2: does the `WakewordTrigger()` take a path to a folder containing the three wakeword models? If so, maybe say 
```suggestion
        WakewordTrigger("path_to_tflite_model_folder"),
```

Author: I'll do what I did for NLU here.

        SpeechRecognizer("spokestack_id", "spokestack_secret"),
    ],
)

pipeline.start()
```

## From text to meaning

> If the event is `recognize`, the raw text of what the user said can be found in `context.transcript`. Translating the text into an action is the job of the [Natural Language Understanding (NLU)](/docs/Concepts/nlu) component. A great thing about Spokestack NLU models is that they run entirely on device. The NLU can be initialized like this:

Reviewer: ```suggestion
If the event is `recognize`, the raw text of what the user said can be found in `context.transcript`. Translating the text into an action is the job of the [Natural Language Understanding (NLU)](/docs/concepts/nlu) component. A great thing about Spokestack NLU models is that they run entirely on device. The NLU can be initialized like this:
```

Reviewer: We haven't talked about events yet, so it's abrupt to say "the event"—what event, and from where?

```python
from spokestack.nlu.tflite import TFLiteNLU

> nlu = TFLiteNLU("path_to_tflite_model")

Reviewer 2: same question as above. is this a folder containing the nlu model + vocab + metadata? if so 
```suggestion
nlu = TFLiteNLU("path_to_tflite_model_folder")
```

Author: It's in the NLU part, but I'll make it both places so you don't need both of them.

```

Some useful links for configuring Spokestack's NLU:

> - [high-level NLU guide](/docs/Concepts/nlu)

Reviewer: ```suggestion
- [high-level NLU guide](/docs/concepts/nlu)
```

- [Python NLU module documentation](nlu)

> - [converting an Alexa or Dialogflow NLU model](/docs/Concepts/export)

Reviewer: ```suggestion
- [converting an Alexa or Dialogflow NLU model](/docs/concepts/export)
- [creating an NLU model from scratch](/docs/concepts/nlu-training-data)
```
I see where you got this list from...I'll make a mental note that the Android guide could use this same addition.

## Talking back to your users

> If you want the full smart speaker experience, you will need to give your application a voice. This can be achieved with text-to-speech (TTS). For more information on TTS, see the [TTS concept guide](/docs/Concepts/tts). TTS playback uses `PyAudioOutput` class which plays audio with the default speaker for the device. Like the NLU, TTS can be used in a callback. Take a look at the example below, which simply speaks what the ASR heard.

Reviewer: ```suggestion
If you want the full smart speaker experience, you will need to give your application a voice. This can be achieved with text-to-speech (TTS). For more information on TTS, see the [TTS concept guide](/docs/concepts/tts). TTS playback uses `PyAudioOutput` class, which plays audio with the default speaker for the device. Like NLU, TTS can be used in a callback. Take a look at the example below, which simply speaks what the ASR heard.
```

```python
@pipeline.event
def on_recognize(context):

>     tts.synthesize("welcome to spokestack")

Reviewer 2: Forgive my ignorance here, will `synthesize` perform automatic playback? If so, is there a way to just get the synthesis file itself?

Author: Synthesize from the TTSManger class does automatic playback. We did add the url option though so I missed that.

### Setup

> We use `pyenv` for virtual environments.

Reviewer: ```suggestion
We use [`pyenv`](https://github.com/pyenv/pyenv) for virtual environments.
```

Might as well...I should've suggested this earlier.

## Integration

In order for your application to use Spokestack's features, there are a few things you will need:

- A free Spokestack Account
- Audio Input Device
- A `SpeechPipeline` Instance
- Audio Output Device

### 1. Spokestack Account

Go to [spokestack.io](/create) to set up your own account (it's free!). Once you've got that, go [grab one of our free NLU models](/account/services/nlu). We'll use the `Highlow` one in this example, but you can choose another, or [create your own](/docs/concepts/nlu-training-data)

Once you've downloaded your NLU, unzip `nlu.tar.gz` with the three files inside (`metadata.json`, `nlu.tflite`, `vocab.txt`). The location of the directory isn't important, because we will pass the path on initialization.

### 2. Audio Input Device

The `PyAudioInput` class will use the system default audio input device. Most personal computers have some form of microphone, but in the case of an embedded device, you may need to purchase a small USB microphone.

### 3. `SpeechPipeline` Instance

> The `SpeechPipeline` [guide](speech-pipeline) has a detailed explanation of how to set up the pipeline, so we will show the quickest way here using a profile.

Reviewer: We haven't explained what the pipeline or profiles are at this point, so maybe a quick introduction for each...

```suggestion
Spokestack's speech pipeline handles collecting audio from the input device and transcribing speech directed at your app. The [`SpeechPipeline` guide](speech-pipeline) has a detailed explanation of how to set up the pipeline, so we will show the quickest way here using a profile, which configures the pipeline's components for a specific use case. The profile we use here includes wakeword activation and speech transcription using Spokestack's cloud ASR.
```

```python
from spokestack.profile.wakeword_asr import WakewordSpokestackASR

pipeline = WakewordSpokestackASR.create(
    "spokestack_id", "spokestack_secret", model_dir="path_to_tflite_model_dir"
)
pipeline.start()
```

## From text to meaning

Translating the text into an action is the job of the [Natural Language Understanding (NLU)](/docs/concepts/nlu) component. A great thing about Spokestack NLU models is that they run entirely on device. The NLU can be initialized like this:

```python
from spokestack.nlu.tflite import TFLiteNLU

nlu = TFLiteNLU("path_to_tflite_model_dir")
```

> Input to the NLU model is the ASR transcript. The transcript can be accessed as a property of `SpeechContext`. Below is a sample Event Handler for running inference on the speech transcript.

Reviewer: ```suggestion
Input to the NLU model is the ASR transcript. The transcript can be accessed as a property of `SpeechContext`. Below is a sample event handler for running inference on the speech transcript.
```

```python
@pipeline.event
def on_recognize(context)
    results = nlu(context.transcript)
```

Some useful links for configuring Spokestack's NLU:

- [high-level NLU guide](/docs/concepts/nlu)
- [Python NLU module documentation](nlu)
- [converting an Alexa or Dialogflow NLU model](/docs/concepts/export)

## Talking back to your users

> If you want the full smart speaker experience, you will need to give your application a voice. This can be achieved with text-to-speech (TTS). For more information on TTS, see the [TTS concept guide](/docs/concepts/tts). TTS playback uses `PyAudioOutput` class, which plays audio with the default speaker for the device. Like NLU, TTS can be used in an Event Handler. Take a look at the example below, which simply speaks what the ASR heard.

Reviewer: ```suggestion
If you want the full smart speaker experience, you will need to give your application a voice. This can be achieved with text-to-speech (TTS). For more information on TTS, see the [TTS concept guide](/docs/concepts/tts). TTS playback uses the `PyAudioOutput` class, which plays audio with the default speaker for the device. Like NLU, TTS can be used in an event handler. Take a look at the example below, which simply speaks what the ASR heard.
```
